File: backend/server/routes/answers.ts
// Answer API Routes
import express from "express";
import { isAuthenticated } from "../auth";
import { storage } from "../index";
import { insertAnswerSchema, toUUID } from "@shared/schema";

const router = express.Router();

// GET /api/answers - Get answers for authenticated user
router.get("/", isAuthenticated, async (req: any, res) => {
  try {
    const userId = req.user.id;
    const { interviewId, questionId, competency } = req.query;
    
    let answers;
    if (interviewId) {
      // Get answers for specific interview
      answers = await storage.getAnswersByInterviewId(toUUID(interviewId as string));
      
      // Verify interview belongs to authenticated user
      const interview = await storage.getInterviewById(toUUID(interviewId as string));
      if (!interview || interview.userId !== userId) {
        return res.status(403).json({ message: "Access denied" });
      }
    } else if (questionId) {
      // Get answers for specific question
      answers = await storage.getAnswersByQuestionId(toUUID(questionId as string));
      // TODO: Add proper authorization check for question ownership
    } else if (competency) {
      answers = await storage.getAnswersByCompetency(userId, competency as string);
    } else {
      // Get all answers for user (across all interviews)
      const userInterviews = await storage.getInterviewsByUserId(userId);
      const allAnswers = await Promise.all(
        userInterviews.map(interview => storage.getAnswersByInterviewId(toUUID(interview.id)))
      );
      answers = allAnswers.flat();
    }
    
    res.json(answers);
  } catch (error) {
    console.error("Get answers error:", error);
    res.status(500).json({ message: error instanceof Error ? error.message : "Unknown error" });
  }
});

// GET /api/answers/by-interview/:interviewId - Get answers for specific interview  
router.get("/by-interview/:interviewId", isAuthenticated, async (req: any, res) => {
  try {
    const { interviewId } = req.params;
    const userId = req.user.id;
    
    console.log("Getting answers for interview:", interviewId, "user:", userId);
    
    // Verify interview belongs to authenticated user
    const interview = await storage.getInterviewById(toUUID(interviewId));
    if (!interview) {
      return res.status(404).json({ message: "Interview not found" });
    }
    
    if (interview.userId !== userId) {
      return res.status(403).json({ message: "Access denied" });
    }
    
    // Get answers for this interview
    const answers = await storage.getAnswersByInterviewId(toUUID(interviewId));
    
    console.log("Found answers for interview:", answers.length);
    res.json(answers);
  } catch (error) {
    console.error("Get answers for interview error:", error);
    res.status(500).json({ message: error instanceof Error ? error.message : "Unknown error" });
  }
});

// GET /api/answers/by-question/:questionId - Get answers for specific question (latest first)
router.get("/by-question/:questionId", isAuthenticated, async (req: any, res) => {
  try {
    const { questionId } = req.params;
    const userId = req.user.id;
    
    // Validate UUID format
    const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
    if (!uuidRegex.test(questionId)) {
      return res.status(400).json({ message: "Invalid question ID format" });
    }
    
    // Verify question exists and belongs to user's interview
    const question = await storage.getQuestionById(toUUID(questionId));
    if (!question) {
      return res.status(404).json({ message: "Question not found" });
    }
    
    const interview = await storage.getInterviewById(toUUID(question.interviewId!));
    if (!interview || interview.userId !== userId) {
      return res.status(403).json({ message: "Access denied" });
    }
    
    // Get answers for this question (sorted by creation date DESC)
    const answers = await storage.getAnswersByQuestionId(toUUID(questionId));
    res.json(answers);
  } catch (error) {
    console.error("Get answers by question error:", error);
    res.status(500).json({ message: error instanceof Error ? error.message : "Unknown error" });
  }
});

// GET /api/answers/:id - Get specific answer
router.get("/:id", isAuthenticated, async (req: any, res) => {
  try {
    const { id } = req.params;
    const userId = req.user.id;
    const answerId = id; // UUID, no need to parse
    
    // Validate UUID format
    const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
    if (!uuidRegex.test(answerId)) {
      return res.status(400).json({ message: "Invalid answer ID format" });
    }
    
    const answer = await storage.getAnswerById(toUUID(answerId));
    if (!answer) {
      return res.status(404).json({ message: "Answer not found" });
    }
    
    // Verify answer belongs to authenticated user by checking session ownership
    const interview = await storage.getInterviewById(toUUID(answer.interviewId!));
    if (!interview || interview.userId !== userId) {
      return res.status(403).json({ message: "Access denied" });
    }
    
    res.json(answer);
  } catch (error) {
    console.error("Get answer error:", error);
    res.status(500).json({ message: error instanceof Error ? error.message : "Unknown error" });
  }
});

// POST /api/answers - Create new answer
router.post("/", isAuthenticated, async (req: any, res) => {
  try {
    const userId = req.user.id;
    const { interviewId, questionId, answerText, timeSpent } = req.body;
    
    // Verify session belongs to authenticated user
    const interview = await storage.getInterviewById(interviewId);
    if (!interview || interview.userId !== userId) {
      return res.status(403).json({ message: "Access denied to interview session" });
    }
    
    // Verify question exists and belongs to the session
    const question = await storage.getQuestionById(questionId);
    if (!question || question.interviewId !== interviewId) {
      return res.status(400).json({ message: "Invalid question for this session" });
    }
    
    // CRITICAL FIX: Check if answer already exists for this question
    const existingAnswers = await storage.getAnswersByQuestionId(toUUID(questionId));
    let answer;
    
    if (existingAnswers.length > 0) {
      // Update existing answer instead of creating duplicate
      const existingAnswer = existingAnswers[0];
      const answerData = {
        answerText,
        timeSpent: timeSpent || 0
      };
      answer = await storage.updateAnswer(toUUID(existingAnswer.id), answerData);
      console.log('Updated existing answer for question:', questionId);
    } else {
      // Create new answer
      const answerData = insertAnswerSchema.parse({
        interviewId,
        questionId,
        answerText,
        timeSpent: timeSpent || 0
      });
      answer = await storage.createAnswer(answerData);
      console.log('Created new answer for question:', questionId);
    }
    
    if (!answer) {
      throw new Error('Failed to create or update answer');
    }

    // No need to track global answer count - free users are limited per interview

    // Trigger AI evaluation for the answer
    try {
      const { evaluateAnswer } = await import("../services/openai.js");
      
      // Get CV context if available
      let cvContext = null;
      try {
        const cvDoc = await storage.getDocumentByType(userId, "cv");
        cvContext = cvDoc?.content || null;
      } catch (error) {
        console.log("No CV found for evaluation context");
      }

      const evaluation = await evaluateAnswer(
        question.questionText,
        answerText,
        question.competency,
        cvContext || undefined
      );

      // Debug the evaluation object structure
      console.log('Full evaluation object:', JSON.stringify(evaluation, null, 2));

      // Store the rating - ensure proper data types for database
      const ratingData = {
        answerId: answer.id,
        overallScore: evaluation.overallScore.toString(), // Convert to string for decimal type
        competencyScores: evaluation.competencyScores, // Already an object for JSONB
        starMethodAnalysis: evaluation.starMethodAnalysis, // Already an object for JSONB
        feedback: evaluation.feedback,
        strengths: evaluation.strengths, // Keep as array for JSONB
        improvementAreas: evaluation.improvementAreas, // Keep as array for JSONB
        aiImprovedAnswer: evaluation.improvedAnswer
      };
      
      console.log('Rating data being stored:', JSON.stringify(ratingData, null, 2));

      try {
        const rating = await storage.createRating(ratingData);
        console.log('Rating successfully created:', JSON.stringify(rating, null, 2));
      } catch (ratingError) {
        console.error('Failed to create rating:', ratingError);
        console.error('Rating data that failed:', JSON.stringify(ratingData, null, 2));
        // Continue without failing the entire request
      }

      // Return answer with evaluation
      res.status(201).json({
        ...answer,
        evaluation
      });
    } catch (evaluationError) {
      console.error("AI evaluation failed:", evaluationError);
      // Still return the answer even if evaluation fails
      res.status(201).json(answer);
    }
  } catch (error) {
    console.error("Create answer error:", error);
    res.status(500).json({ message: error instanceof Error ? error.message : "Unknown error" });
  }
});

// PATCH /api/answers/:id - Update answer
router.patch("/:id", isAuthenticated, async (req: any, res) => {
  try {
    const { id } = req.params;
    const userId = req.user.id;
    const answerId = id; // UUID, no need to parse
    
    // Validate UUID format
    const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
    if (!uuidRegex.test(answerId)) {
      return res.status(400).json({ message: "Invalid answer ID format" });
    }
    
    const answer = await storage.getAnswerById(toUUID(answerId));
    if (!answer) {
      return res.status(404).json({ message: "Answer not found" });
    }
    
    // Verify answer belongs to authenticated user by checking session ownership
    const interview = await storage.getInterviewById(toUUID(answer.interviewId!));
    if (!interview || interview.userId !== userId) {
      return res.status(403).json({ message: "Access denied" });
    }
    
    const updatedAnswer = await storage.updateAnswer(toUUID(answerId), req.body);
    res.json(updatedAnswer);
  } catch (error) {
    console.error("Update answer error:", error);
    res.status(500).json({ message: error instanceof Error ? error.message : "Unknown error" });
  }
});

// DELETE /api/answers/:id - Delete answer
router.delete("/:id", isAuthenticated, async (req: any, res) => {
  try {
    const { id } = req.params;
    const userId = req.user.id;
    const answerId = id; // UUID, no need to parse
    
    // Validate UUID format
    const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
    if (!uuidRegex.test(answerId)) {
      return res.status(400).json({ message: "Invalid answer ID format" });
    }
    
    const answer = await storage.getAnswerById(toUUID(answerId));
    if (!answer) {
      return res.status(404).json({ message: "Answer not found" });
    }
    
    // Verify answer belongs to authenticated user by checking session ownership
    const interview = await storage.getInterviewById(toUUID(answer.interviewId!));
    if (!interview || interview.userId !== userId) {
      return res.status(403).json({ message: "Access denied" });
    }
    
    // TODO: Implement answer deletion with cascade to ratings
    res.json({ message: "Answer deleted successfully" });
  } catch (error) {
    console.error("Delete answer error:", error);
    res.status(500).json({ message: error instanceof Error ? error.message : "Unknown error" });
  }
});

export default router;

File: python-backend/src/api/routes/answers.py
"""Answer API routes.

Handles HTTP requests for answer operations.
"""

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query

from src.api.dependencies import DbSession
from src.api.middleware.auth import CurrentUser
from src.application.dto.answer import (
    AnswerCreateRequest,
    AnswerResponse,
    AnswerUpdateRequest,
    AnswerWithEvaluationResponse,
    MessageResponse,
)
from src.application.services.answer_service import AnswerService
from src.domain.entities import Answer
from src.infrastructure.repositories.answer_repository import AnswerRepository
from src.infrastructure.repositories.interview_repository import InterviewRepository
from src.infrastructure.repositories.question_repository import QuestionRepository

router = APIRouter(prefix="/api/answers", tags=["answers"])


def get_answer_service(db: DbSession) -> AnswerService:
    """Create answer service with dependencies.

    Args:
        db: Database session from DI

    Returns:
        Configured AnswerService instance
    """
    answer_repo = AnswerRepository(db)
    interview_repo = InterviewRepository(db)
    question_repo = QuestionRepository(db)
    return AnswerService(answer_repo, interview_repo, question_repo)


AnswerServiceDep = Annotated[AnswerService, Depends(get_answer_service)]


def _to_response(answer: Answer) -> AnswerResponse:
    """Convert Answer entity to response DTO."""
    return AnswerResponse(
        id=answer.id,
        interviewId=answer.interview_id,
        questionId=answer.question_id,
        answerText=answer.answer_text,
        timeSpent=answer.time_spent,
        answeredAt=answer.answered_at,
    )


@router.get("")
async def get_answers(
    current_user: CurrentUser,
    answer_service: AnswerServiceDep,
    interview_id: UUID | None = Query(None, alias="interviewId"),
    question_id: UUID | None = Query(None, alias="questionId"),
    competency: str | None = Query(None),
) -> list[AnswerResponse]:
    """Get answers filtered by interviewId, questionId, or competency."""
    try:
        if interview_id:
            answers = await answer_service.get_answers_by_interview(
                interview_id, current_user.id
            )
        elif question_id:
            answers = await answer_service.get_answers_by_question(
                question_id, current_user.id
            )
        elif competency:
            answers = await answer_service.get_answers_by_competency(
                competency, current_user.id
            )
        else:
            # Get all answers across all interviews
            # This is handled by getting by competency with no filter
            # but that requires iterating all interviews
            raise HTTPException(
                status_code=400,
                detail="interviewId, questionId, or competency parameter required",
            )

        return [_to_response(a) for a in answers]

    except PermissionError as e:
        raise HTTPException(status_code=403, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/by-interview/{interview_id}")
async def get_answers_by_interview(
    interview_id: UUID,
    current_user: CurrentUser,
    answer_service: AnswerServiceDep,
) -> list[AnswerResponse]:
    """Get all answers for a specific interview."""
    try:
        answers = await answer_service.get_answers_by_interview(
            interview_id, current_user.id
        )
        return [_to_response(a) for a in answers]

    except PermissionError as e:
        raise HTTPException(status_code=403, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/by-question/{question_id}")
async def get_answers_by_question(
    question_id: UUID,
    current_user: CurrentUser,
    answer_service: AnswerServiceDep,
) -> list[AnswerResponse]:
    """Get answers for a specific question."""
    try:
        answers = await answer_service.get_answers_by_question(
            question_id, current_user.id
        )
        return [_to_response(a) for a in answers]

    except PermissionError as e:
        raise HTTPException(status_code=403, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/{answer_id}")
async def get_answer(
    answer_id: UUID,
    current_user: CurrentUser,
    answer_service: AnswerServiceDep,
) -> AnswerResponse:
    """Get a specific answer by ID."""
    try:
        answer = await answer_service.get_answer(answer_id, current_user.id)
        return _to_response(answer)

    except PermissionError as e:
        raise HTTPException(status_code=403, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("", status_code=201)
async def create_answer(
    request: AnswerCreateRequest,
    current_user: CurrentUser,
    answer_service: AnswerServiceDep,
) -> AnswerWithEvaluationResponse:
    """Create a new answer.

    If an answer already exists for the question, it will be updated.
    AI evaluation will be added in a future update.
    """
    try:
        answer = await answer_service.create_answer(
            interview_id=request.interview_id,
            question_id=request.question_id,
            answer_text=request.answer_text,
            requesting_user_id=current_user.id,
            time_spent=request.time_spent,
        )

        # TODO: Trigger AI evaluation (Task 15-16)
        # For now, return without evaluation
        response = AnswerWithEvaluationResponse(
            id=answer.id,
            interviewId=answer.interview_id,
            questionId=answer.question_id,
            answerText=answer.answer_text,
            timeSpent=answer.time_spent,
            answeredAt=answer.answered_at,
            evaluation=None,
        )

        return response

    except PermissionError as e:
        raise HTTPException(status_code=403, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.patch("/{answer_id}")
async def update_answer(
    answer_id: UUID,
    request: AnswerUpdateRequest,
    current_user: CurrentUser,
    answer_service: AnswerServiceDep,
) -> AnswerResponse:
    """Update an existing answer."""
    try:
        answer = await answer_service.update_answer(
            answer_id=answer_id,
            requesting_user_id=current_user.id,
            answer_text=request.answer_text,
            time_spent=request.time_spent,
        )
        return _to_response(answer)

    except PermissionError as e:
        raise HTTPException(status_code=403, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.delete("/{answer_id}")
async def delete_answer(
    answer_id: UUID,
    current_user: CurrentUser,
    answer_service: AnswerServiceDep,
) -> MessageResponse:
    """Delete an answer."""
    try:
        deleted = await answer_service.delete_answer(answer_id, current_user.id)
        if deleted:
            return MessageResponse(message="Answer deleted successfully")
        raise HTTPException(status_code=500, detail="Failed to delete answer")

    except PermissionError as e:
        raise HTTPException(status_code=403, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
